docs/DataProviders.md

```diff
@@ -650,4 +586,263 @@ const UserProfile = ({ userId }) => {
+
+```js
+import { withLifecycleCallbacks } from 'react-admin';
+import simpleRestProvider from 'ra-data-simple-rest';
+
+const dataProvider = withLifecycleCallbacks(simpleRestProvider('http://path.to.my.api/'), [
+    {
+        /**
+         * For posts update only, convert uploaded images to base 64 and attach them to
+         * the `picture` sent property, with `src` and `title` attributes.
+         */
+        resource: 'posts',
+        beforeUpdate: async (params, dataProvider) => {
+            // Freshly dropped pictures are File objects and must be converted to base64 strings
+            const newPictures = params.data.pictures.filter(
+                p => p.rawFile instanceof File
+            );
+            const formerPictures = params.data.pictures.filter(
+                p => !(p.rawFile instanceof File)
+            );
+
+            const base64Pictures = await Promise.all(
+                newPictures.map(convertFileToBase64)
+            )
+            const pictures = [
+                ...base64Pictures.map((dataUrl, index) => ({
+                    src: dataUrl,
+                    title: newPictures[index].name,
+                })),
+                ...formerPictures,
+            ];
+            return dataProvider.update(
+                resource,
+                { data: { ...params.data, pictures } }
+            );  
+        }
+    }
+]);
+
+/**
+ * Convert a `File` object returned by the upload input into a base 64 string.
+ * That's not the most optimized way to store images in production, but it's
+ * enough to illustrate the idea of dataprovider decoration.
+ */
+const convertFileToBase64 = file =>
+    new Promise((resolve, reject) => {
+        const reader = new FileReader();
+        reader.onload = () => resolve(reader.result);
+        reader.onerror = reject;
+        reader.readAsDataURL(file.rawFile);
+    });
+
+export default myDataProvider;
+```
+
+**Tip**: Use `beforeSave` instead of `beforeUpdate` to do the same for both create and update calls.
+
+### Sending Files In `multipart/form-data`
+
+In case you need to upload files, as you would with an HTML form but to your API, you can use the [FormData](https://developer.mozilla.org/en-US/docs/Web/API/FormData) API . It uses the same format as a form would use if the encoding type were set to `multipart/form-data`.
+
+The following `dataProvider` extends an existing one and tweaks its `create` and `update` methods for the `posts` resource only. It is the role of your API to negotiate the request and process the image. 
+
+The data provider sends the `post` data with the attached file and works this way:
+- it checks if the resource is `posts`;
+- if so, it creates a new `FormData` object with the file received from the form and it sends this `FormData` to the API with the react-admin `fetchUtils.fetchJson()` function (a `fetch().then(r => r.json())` shortcut); 
+- if not, it simply uses the `baseDataProvider` base methods.
```

Comment:
These explanations are a bit confusing to me because:

- A lot of it is dedicated to explaining when to trigger the FormData implementation, and what to do when it is not triggered ; but the interesting part is actually the rest, not this
- you can probably skip all these details, and focus on the interesting part, by rewriting the example below to use `withLifecycleCallbacks`, and simply linking to its documentation for people not yet knowing how it works
- it is not clear enough (IMHO) that the whole form values will be sent inside the `FormData` object, not only the image file
- The use of `fetchUtils` is an implementation detail, you could as well do it with a simple `fetch`. I'd either not mention it at all, or simply mention it and provide a link to its doc, without explaining further what it does, as it adds unnecessary complexity IMHO